### PR TITLE
Switch CpsDiagnosticItemSourceProvider to use IProjectTree/ProjectTreeFlags

### PR DESCRIPTION
**Pull request by ToddGrun**
This shows up in both allocations/CPU during solution explorer search locally.

### CODE REVIEW

**Comment by ToddGrun** [comment]
Going to measure this afternoon

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs [hunk 1]
```diff
@@ -100,66 +101,61 @@ internal sealed class CpsDiagnosticItemSourceProvider : AttachedCollectionSource
...
             {
-                var capabilitiesString = (string)capabilitiesObj;
+                if (item.HierarchyIdentity.NestedHierarchy is IVsBrowseObjectContext { UnconfiguredProject.Services.ProjectTreeService.CurrentTree.Tree: { } tree })
+                {
+                    flags = tree.Flags;
+                    return true;
+                }
+            }
+            else
+            {
```
Comment (drewnoakes) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:135]: Do we need this block? Looks to me that if the item is the root, the `while` loop below will behave the same as the `if`. I think they can be merged into one block, and the code de-duplicated.

Comment (drewnoakes) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:135]: If I'm wrong, then the comment should still be hoisted up to above the `if`, as it applies to both blocks.

Comment (ToddGrun) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:135]: If I didn't do this, then if the passed in item was a root, the TryFind call would end up failing.

I don't object to moving the comment, but it seems most applicable here as this is the code that walks up the tree to find the appropriate object on which to create the IVsBrowseObjectContext

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs [hunk 2]
```diff
@@ -100,66 +101,61 @@ internal sealed class CpsDiagnosticItemSourceProvider : AttachedCollectionSource
         /// </summary>
         private static string? GetTargetFrameworkMoniker(IVsHierarchyItem item)
```
Comment (drewnoakes) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:102]: Unrelated to this PR, but the returned string here is actually the target framework "alias", not the moniker.

The moniker looks like `.NETFramework,Version=v4.5.1` while the alias looks like `net451`.

Unfortunately the TFM prefix on the flags is misleading, but it's too late to change that now.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs (hunk 1 above)
Comment (drewnoakes) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs:135]: Ok I see. The root will never be part of the dependencies tree, so for this usage it's fine to return no flags for the root. But no issue to leave as-is. If you want to keep it, maybe add a comment that `TryFind(IntPtr)` doesn't work for the root node, as it's unclear why it's given special treatment.

**Review by drewnoakes — APPROVED** [review]

**Comment by ToddGrun** [comment]
> Going to measure this afternoon

Generally looks good.

OLD MEMORY:
image

NEW MEMORY:
image

OLD CPU:
image

NEW CPU:
image

**Comment by lifengl** [comment]
looks good to me.

**Review by lifengl — APPROVED** [review]
looks good to me

**Comment by ToddGrun** [comment]
ptal @jasonmalinowski or @genlu

**Review by jasonmalinowski — APPROVED** [review]
As long as project system folks are signing off, I'm good with this. :smile:

File: src/VisualStudio/Core/Impl/Microsoft.VisualStudio.LanguageServices.Implementation.csproj [hunk 3]
```diff
@@ -39,6 +39,7 @@
     <Reference Include="WindowsFormsIntegration" />
     <EmbeddedResource Update="SolutionExplorerShim.resx" GenerateSource="true" />
     <PackageReference Include="Microsoft.CSharp" />
+    <PackageReference Include="Microsoft.VisualStudio.ProjectSystem" />
```
Comment (drewnoakes) [inline src/VisualStudio/Core/Impl/Microsoft.VisualStudio.LanguageServices.Implementation.csproj:42]: @lifengl would it be better to use the SDK package here, to get CPS's analyzers etc?

**Pull request merged**